src/Broadcast.chpl

```diff
@@ -135,4 +145,36 @@ module Broadcast {
     expandedVals = (+ scan expandedVals);
     return (expandedVals == 1);
   }
+
+  proc broadcast(segs: [?sD] int, segString: borrowed SegString, size: int) throws {
+    ref segOff = segString.offsets.a;
+    const high = sD.high;
+    var strSize: int;
+    var diffs: [sD] int;
+
+    forall (i, d, seg, off) in zip (sD, diffs, segs, segOff) with (+ reduce strSize) {
+      if i == high {
+        strSize += (size - seg) * (segString.nBytes - off);
+        d = segString.nBytes - off;
+      } else {
+        strSize += (segs[i+1] - seg) * (segOff[i+1] - off);
+        d = segOff[i+1] - off;
+      }
+    }
+
+    var broadDist = broadcast(segs, diffs, size);
+    var offsets = (+ scan broadDist) - broadDist;
+    var r: [sD] int = 0..segString.size; 
+    var ind = broadcast(segs, r, size);
+    var expandedVals = makeDistArray(strSize, uint(8));
+    ref vals = segString.values.a;
+
+    forall (i, o, s) in zip(ind, offsets, 0..#size) with (ref vals) {
+      var inds = if i == high then segOff[i]..segString.nBytes-1 else segOff[i]..segOff[i+1]-1;
+      var offs = if s == size - 1 then o..strSize-1 else o..offsets[s+1]-1;
+      expandedVals[offs] = vals[inds];
+    }
```

Comment:
im curious if we can make this a bit more similar to other strings workflows and take a localizing slice here and maybe do some aggregation. I think the advantage this buys us is to ensure the `vals` are local

Right now we calculate the ranges for `inds` and `offs`, but they both have to be the same length. So we really just need the starts of each and how long the str is. I think something like this might give us better perf. (warning this is psuedocode, i've not run it :)
```chapel
const lengths = segString.getLengths();
forall (so, o, len) in zip(segOff, offsets, lengths) with (var valAgg = newDstAggregator(uint(8))) {
  var localizedVals = new lowLevelLocalizingSlice(vals, so..#len);
  for k in 0..#len {
    valAgg.copy(expandedVals[o+k], localizedVals.ptr[k]);
  }
}
```

im not sure this is gonna be quicker or not, we should do some comm diags to compare. I'm also not sure if it's within the scope of this PR or not, but def wanted to make a note of it before i forgot

Reply:
so i ran just some real quick perf numbers (super small scale like 100 long arrays) and i am seeing some lowered comm using the above
```python
In [15]: a = ak.arange(10)

In [16]: segs = ak.arange(0,100,10)

In [17]: a
Out[17]: array([0 1 2 3 4 5 6 7 8 9])

In [18]: segs
Out[18]: array([0 10 20 30 40 50 60 70 80 90])

In [19]: ak.broadcast(segs, a, 100)
Out[19]: array([0 0 0 ... 9 9 9])

In [20]: s = ak.cast(a, str)

In [21]: s
Out[21]: array(['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'])

In [22]: ak.broadcast(segs, s, 100)
Out[22]: array(['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '1', '1', '1', '1', '1', '1', '1', '1', '1', '1', '2', '2', '2', '2', '2', '2', '2', '2', '2', '2', '3', '3', '3', '3', '3', '3', '3', '3', '3', '3', '4', '4', '4', '4', '4', '4', '4', '4', '4', '4', '5', '5', '5', '5', '5', '5', '5', '5', '5', '5', '6', '6', '6', '6', '6', '6', '6', '6', '6', '6', '7', '7', '7', '7', '7', '7', '7', '7', '7', '7', '8', '8', '8', '8', '8', '8', '8', '8', '8', '8', '9', '9', '9', '9', '9', '9', '9', '9', '9', '9'])
```
and the comm diagnostics for this
```
USING whats there rn

| locale | get | put | execute_on | execute_on_nb |
| -----: | --: | --: | ---------: | ------------: |
|      0 | 818 | 300 |          0 |           601 |
|      1 | 811 | 100 |        200 |           600 |
USING agg and localizing slices

| locale | get | put | execute_on | execute_on_nb |
| -----: | --: | --: | ---------: | ------------: |
|      0 |   1 |   5 |          0 |             4 |
|      1 |  32 |   6 |         10 |             0 |
```

Reply:
posted a comment about comm, but mighta spoken too early the impl above is only writing the first couple vals, looking into it more

Reply:
Digging a bit more, there's def perf improvements to be gained from an approach like this

I had a `10**4` long strings and ran both methods and recorded the comm and time:
```python
In [3]: def test(i):
   ...:     TOT = 10**i
   ...:     SIZE = 10**(i-1)
   ...:     a = ak.arange(SIZE)
   ...:     segs = ak.arange(0,TOT,10)
   ...:     s = ak.cast(a, str)
   ...:     rand = ak.random_strings_uniform(1, 16, SIZE)
   ...:     s = s.stick(rand, delimiter="_")
   ...:     return ak.broadcast(segs, s, TOT)
   ...: 

In [4]: test(5)
Out[4]: array(['0_WXVTUR', '0_WXVTUR', '0_WXVTUR', ... , '9999_OZH', '9999_OZH', '9999_OZH'])
```
this is the results:
```
new way time = 0

| locale | get | put | execute_on | execute_on_nb |
| -----: | --: | --: | ---------: | ------------: |
|      0 |  17 |   1 |          0 |            41 |
|      1 | 177 |  13 |          5 |             0 |
old way time = 115

| locale |    get |    put | execute_on | execute_on_nb |
| -----: | -----: | -----: | ---------: | ------------: |
|      0 | 800064 | 300000 |          0 |        600063 |
|      1 | 800203 | 100017 |     200004 |        600378 |
```

the code i suggest using is:
```chapel
proc broadcast(segs: [?sD] int, segString: borrowed SegString, size: int) throws {
  ref offs = segString.offsets.a;
  ref vals = segString.values.a;
  const high = sD.high;
  var strLens: [sD] int;
  var segLens: [sD] int;
  var expandedLen: int;

  forall (i, off, str_len, seg, seg_len) in zip (sD, offs, strLens, segs, segLens) with (+ reduce expandedLen) {
    if i == high {
      seg_len = size - seg;
      str_len = vals.size - off;
    } else {
      seg_len = segs[i+1] - seg;
      str_len = offs[i+1] - off;
    }
    expandedLen += seg_len * str_len;
  }
  var broadDist = broadcast(segs, strLens, size);
  const offsets = (+ scan broadDist) - broadDist;
  var expandedVals = makeDistArray(expandedLen, uint(8));

  forall (off, str_len, seg, seg_len) in zip(offs, strLens, segs, segLens) with (var valAgg = newDstAggregator(uint(8))) {
    var localizedVals = new lowLevelLocalizingSlice(vals, off..#str_len);
    for i in seg..#seg_len {
      var expValOff = offsets[i];
      for k in 0..#str_len {
        valAgg.copy(expandedVals[expValOff+k], localizedVals.ptr[k]);
      }
    }
  }
  return (expandedVals, offsets);
}
```